### PR TITLE
Make clone dispatch to `clone` method if present.

### PR DESCRIPTION
**Pull request by nomosyn**
Fixes: #1292

### CODE REVIEW

File: test/clone.js [hunk 1]
```diff
@@ -189,3 +189,20 @@ describe('deep clone edge cases', function() {
...
 });
+
+
+describe('Let `R.clone` use an arbitrary user defined `clone` method', function() {
+
+  it('dispatches to `clone` method if present', function() {
+    function ArbitraryClone(x) { this.value = x; }
+
+    // user defined `clone` method can or cannot make sense: it's arbitrary.
+    ArbitraryClone.prototype.clone = function() { return new ArbitraryClone(2 * this.value); };
```
Comment (CrossEye) [inline test/clone.js]: Although clearly an `ArbitraryClone` would be allowed to do whatever it wanted to, I'd rather _not_ call attention to it in the test suite, just doing, `return new ArbitraryClone(this.value)`.

If you find this that important, I think it should just be a comment.  ("Note that if Ramda delegates to the object's 'clone' method, it does no checking that data is similar," or some such.)  But personally, I don't find it really worth spending any effort on.

**Comment by CrossEye** [comment]
Thank you.  Except for the minor issue mentioned about the new test, this looks good to me.

**Comment by nomosyn** [comment]
Ok! Updated.

**Comment by buzzdecafe** [comment]
:cow2:

**Comment by nomosyn** [comment]
mmh... That's a cow.

**Comment by davidchambers** [comment]
Does the special case for arrays in `_hasMethod` both anyone else in this context? I expect `let x = []; x.clone = () => 42; let y = R.clone(x);` to bind `42` to `y`. The fact that it doesn't suggests an implementation detail is affecting the API.

**Comment by CrossEye** [comment]
> Does the special case for arrays in `_hasMethod` both anyone else in this context? 

No more than it does for `chain` or `of` or `tail`.  It seems slightly unfortunate, but not tragic.  If you extend an array, or `Array.prototype` with a new method, then our dispatch mechanism might well skip it when trying to convert your API from `obj.method(...params)` to `function(...params, obj)`.  But your original sytax will still work.  That seems as far as it goes.
